dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -189,16 +230,38 @@ private UfsUrl(UfsUrlMessage proto) {
 
   /**
    * Constructs an {@link UfsUrl} from components.
-   * @param scheme the scheme of the path
+   *
+   * @param scheme    the scheme of the path
    * @param authority the authority of the path
-   * @param path the path component of the UfsUrl
+   * @param path      the path component of the UfsUrl
    */
   public UfsUrl(String scheme, String authority, String path) {
-    Preconditions.checkArgument(!scheme.isEmpty(), "scheme is not allowed to be empty,"
-        + "please input again.");
+    UfsUrl ufsUrl = new UfsUrl(scheme, authority, path, true);
+    mProto = ufsUrl.toProto();
+  }
+
+  /**
+   * Constructs an {@link UfsUrl} from components.
+   * Note that if checkNormalization is false, it will also remove redundant slashes of path.
+   *
+   * @param scheme    the scheme of the path
+   * @param authority the authority of the path
+   * @param path      the path component of the UfsUrl
+   * @param checkNormalization  the parameter to decide whether normalize the path string
+   */
+  public UfsUrl(String scheme, String authority, String path, boolean checkNormalization) {
+    Preconditions.checkArgument(!scheme.isEmpty(), "empty scheme: %s",
+        scheme + authority + path);
     Preconditions.checkArgument(!scheme.equalsIgnoreCase("alluxio"),
         "Alluxio 3.x no longer supports alluxio:// scheme,"
             + " please input the UFS path directly like hdfs://host:port/path");
```

Comment:
also check if the scheme is valid, e.g. does not itself contain `/` or `:`. 

If you don't want to do the checking, then restrict the ways a caller can use to construct an instance. Either require callers to use the parser, or use a dedicated type for scheme, and don't just accept unsanitized `String`s as inputs.

Reply:
OK